IdentityCore/src/oauth2/MSIDTokenResponseHandler.m

```diff
@@ -0,0 +1,172 @@
+                                     MSIDErrorInternal, @"processTokenResponse called without a response dictionary", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    [self.class checkCorrelationId:response requestCorrelationId:context.correlationId];
+    
+    if (response.error)
+    {
+        if (error)
+        {
+            MSIDErrorCode errorCode = refreshToken ? MSIDErrorServerRefreshTokenRejected : MSIDErrorServerOauth;
+            if ([response isKindOfClass:MSIDAADV2TokenResponse.class])
+            {
+                errorCode = [self.class getErrorCodeForAADV2:response.error];
+            }
+            
+            *error = MSIDCreateError(MSIDOAuthErrorDomain,
+                                     errorCode,
+                                     response.errorDescription,
+                                     response.error,
+                                     nil,
+                                     nil,
+                                     context.correlationId,
+                                     nil);
+        }
+        return NO;
+    }
+    
+    if (!response.accessToken)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Authentication response received without expected accessToken", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    if (!response.clientInfo)
+    {
+        MSID_LOG_ERROR(context, @"Client info was not returned in the server response");
+        MSID_LOG_ERROR_PII(context, @"Client info was not returned in the server response");
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInternal, @"Client info was not returned in the server response", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    // Checking for V1 response is done
+    if (![response isKindOfClass:MSIDAADV2TokenResponse.class] || ![parameters isKindOfClass:MSIDAADV2RequestParameters.class])
+    {
+        return YES;
+    }
+    
+    // The rest of the checking is only for V2
+    MSIDAADV2RequestParameters *aadV2Parameters = (MSIDAADV2RequestParameters *)parameters;
+    if ([NSString msidIsStringNilOrBlank:response.scope])
+    {
+        MSID_LOG_INFO(context, @"No scope in server response, using passed in scope instead.");
+        MSID_LOG_INFO_PII(context, @"No scope in server response, using passed in scope instead.");
+        [response setScope:aadV2Parameters.scopes.msidToString];
+    }
+    
+    // For silent flow, with grant type being MSID_OAUTH2_REFRESH_TOKEN, this value may be missing from the response.
+    // In this case, we simply return the refresh token in the request.
+    if (refreshToken)
+    {
+        if (!response.refreshToken)
+        {
+            response.refreshToken = refreshToken;
```

Comment:
we shouldn't replace refresh token in the response here. If refresh token was missing, we need to make sure we don't rewrite the previous existing one that we used instead. Are there any cases where this wouldn't work?

Reply:
As discussed, we shouldn't rewrite previous existing token in this case.
And also this logic should be moved to token item class.